### PR TITLE
Features/62 devices

### PR DESCRIPTION
**Pull request by Markus-Goetz**
Adds the possibility to use different devices. So far CPU and GPU are supported.

Untested, need feedback on how we want to proceed with testing this on a platform that does not have any GPUs.

### CODE REVIEW

File: heat/core/devices.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        Retrieves the index of the GPU to be used for this MPI process.
+
+        Returns
+        -------
+        gpu_index : int
+            The GPU index
+        """
+        return 'cuda:{}'.format(MPI_WORLD.rank % torch.cuda.device_count())
+
+    __device_mapping['gpu'] = gpu_index
```
Comment (d1saster) [inline heat/core/devices.py]: I do not see why we have to store a function object here, since the return value of the latter is not subject to change?

**Comment by d1saster** [comment]
With this branch I also experience a segfault when trying to run kmeans with more than one mpi process. Setting the environment variable `PYTHONFAULTHANDLER=1` I get the following output:
```
Current thread 0x00007f07f1d70700 (most recent call first):
  File "/home/knec_ph/src/heat-2/heat/core/communication.py", line 126 in wrapped
  File "/home/knec_ph/src/heat-2/heat/core/operations.py", line 808 in __reduce_op
  File "/home/knec_ph/src/heat-2/heat/core/operations.py", line 545 in sum
  File "/home/knec_ph/src/heat-2/heat/core/tensor.py", line 596 in sum
  File "/home/knec_ph/src/heat-2/heat/ml/cluster/kmeans.py", line 51 in fit
  File "heat/benchmarks/bench_kmeans.py", line 33 in main
  File "heat/benchmarks/bench_kmeans.py", line 36 in <module>
```
The issue seems to be in __reduce_op in line 808:
```
 x.comm.Allreduce(MPI.IN_PLACE, partial[0], op)
```
Here, we just cannot assume that for a pytorch cuda tensor that partial[0] is a pointer to the respective memory, on which Allreduce can operate in place. This should also be fixed.

File: heat/core/devices.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def get_default_device():
```
[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply heat/core/devices.py]: Is currently not, should be though in sanitize_device. Thanks, will fix

File: heat/core/devices.py (hunk 1 above)
Comment (Markus-Goetz) [reply heat/core/devices.py]: Meant for extensibility. Makes it easier to add new backends without having to touch on big if-else or switch-case function that knows all the backends.

For the GPU backend I can for example dynamically check whether it is available or not.

Should stay as is in my opinion.

File: heat/core/random.py [hunk 3]
```diff
@@ -56,6 +58,8 @@ def randn(*args, split=None, comm=MPI_WORLD):
             [ 1.3365, -1.5212,  1.4159, -0.1671],
             [ 0.1260,  1.2126, -0.0804,  0.0907]])
     """
+    # TODO: make me splitable
+    # TODO: add device capabilities
```
[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply heat/core/random.py]: Good point, needs to be fixed, will look into it.

File: heat/core/tensor.py [hunk 4]
```diff
@@ -21,6 +22,10 @@ def __init__(self, array, gshape, dtype, split, comm):
     def comm(self):
         return self.__comm
 
+    @property
+    def device(self):
+        return self.__array.device
```
[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply heat/core/tensor.py]: What do you propose instead? Do you think a string would be sufficient, e.g. 'cpu', 'gpu' or an enumeration type of some sorts?

File: heat/core/devices.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply heat/core/devices.py]: Fixed! Also: will be a function for users to query current default device

**Comment by Markus-Goetz** [comment]
This above issue with the allreduce will be fixed as soon as feature/9-redistribution is done

File: heat/core/tensor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (d1saster) [reply heat/core/tensor.py]: Either way is fine with me.

**Pull request merged**